README.md
# unrust / uni-app

[![Documentation](https://docs.rs/uni-app/badge.svg)](https://docs.rs/uni-app)
[![crates.io](https://meritbadge.herokuapp.com/uni-app)](https://crates.io/crates/uni-app)

This library is a part of [Unrust](https://github.com/unrust/unrust), a pure rust native/wasm game engine.
This library provides a native/wasm compatibility layer for following components :
* Window creation
* Input (mouse + keyboard)
* File system (local + http[s], ready-only)

## Usage

```toml
[dependencies]
uni-app = "0.4.*"
```

```rust
extern crate uni_app;

fn main() {
    // create the game window (native) or canvas (web)
    let app = uni_app::App::new(uni_app::AppConfig {
        size: (800, 600),
        title: "my game".to_owned(),
        vsync: true,
        show_cursor: true,
        headless: false,
        resizable: true,
        fullscreen: false,
        intercept_close_request: false,
        icon: None,
    });
    // start game loop
    app.run(move |app: &mut uni_app::App| {
        for evt in app.events.borrow().iter() {
            // print on stdout (native) or js console (web)
            uni_app::App::print(format!("{:?}", evt));
            // exit when pressing escape
            match &evt {
                uni_app::AppEvent::KeyUp(ev) if ev.code == ScanCode::Escape => {
                    uni_app::App::exit();
                }
                _ => (),
            }
        }
    });
}
```

## Build

### As web app (wasm32-unknown-unknown)

Install wasm32 target :
```
rustup target install wasm32-unknown-unknown
```
Install [wasm-pack](https://rustwasm.github.io/wasm-pack/installer/)
and [npm](https://www.npmjs.com/get-npm)

Compile the demo with
```
wasm-pack build examples
```
This creates a wasm package in examples/pkg

Run the demo with
```
cd www
npm install
npm run start
```

Open your browser at http://localhost:8080/

### As desktop app (native-opengl)

Run it from the www/ directory to be able to load the test.txt file :

```
cd www && cargo run --example basic --release
```

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

You can contribute to this library through pull requests. If you do so, please update the CHANGELOG.md and CREDITS.md files. If you provide a new feature, consider adding an example as a tutorial/showcase.

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog

## [0.5.0] - 2023-03-30
### Added
- on native target, you can set the window icon with AppConfig.icon

## [0.4.1] - 2023-02-17
### Fixed
- scancodes for F11 and F12 (windows & linux)
- scancode for LAlt (windows)

## [0.4.0] - 2023-01-15
### Added
- filesystem example in basic.rs
- URL support in filesystem API on native targets when enabling the http feature

    This breaks API compatibility as File::is_ready(&self) becomes File::is_ready(&mut self)

## [0.3.1] - 2023-01-13
### Fixed
- fixed wasm target compilation

## [0.3.0] - 2023-01-13
### Changed
- keyboard and mouse API revamped KeyUp/DownEvent.code and MouseButtonEvent.button are now enums

### Fixed
- App:print use println on native target to have the same behavior as wasm

### Added
- support for VirtualKeyCode::Apostrophe on mac

## [0.2.1] - 2022-10-11
### Fixed
- get_param on wasm crashing if the URL had no parameters
- right mouse button context menu was not disabled on web platform

## [0.2.0] - 2022-10-10
### Changed
- upgrade to rust edition 2021
- upgrade to glutin 0.29.1
- replace unmaintained cargo-web/stdweb with wasm_bindgen/web-sys

## [0.1.4] - 2020-03-20
### Added
- support for DroppedFile events via App::get_dropped_files()
- upgrade to glutin 0.23

## [0.1.3] - 2019-12-04
### Added
- App::get_screen_resolution()

## [0.1.2] - 2019-11-29
### Added
- support for text input via winit's ReceivedCharacter event

## [0.1.1] - 2019-11-21
### Changed
- upgrade to stdweb 0.4.14
- upgrade to glutin 0.17
- now compiles on stable rust

### Added
- App::exit()

### Fixed
- arrow and pageup/pagedown keys on linux

## [0.1.0] - 2018-06-24
- initial release

examples/basic.rs
extern crate uni_app;
use uni_app::{App, AppConfig, AppEvent, FileSystem, ScanCode};

#[cfg(target_arch = "wasm32")]
use wasm_bindgen::prelude::*;

#[cfg(target_arch = "wasm32")]
#[wasm_bindgen(start)]
pub fn main_js() -> Result<(), JsValue> {
    main();
    Ok(())
}

fn main() {
    // create the game window (native) or canvas (web)
    let app = App::new(AppConfig {
        size: (800, 600),
        title: "my game".to_owned(),
        vsync: true,
        show_cursor: true,
        headless: false,
        resizable: true,
        fullscreen: false,
        intercept_close_request: false,
        icon: None,
    });
    // loading a file
    // if the "http" feature is enabled, this would also work with an url like "https://raw.githubusercontent.com/unrust/uni-app/master/www/test.txt"
    let mut file = Some(FileSystem::open("test.txt").expect("Could not open the file test.txt.\nIn native mode, run this example in the www/ directory :\ncd www && cargo run --example basic"));
    // start game loop
    App::print("Press Escape to exit");
    app.run(move |app: &mut uni_app::App| {
        if let Some(the_file) = &mut file {
            if the_file.is_ready() {
                // when the file is ready, display its content
                let content = the_file
                    .read_text()
                    .expect("Could not read test.txt file's content");
                App::print(format!("File content : {}", content));
                // to avoid doing it every frame
                file = None;
            }
        }
        for evt in app.events.borrow().iter() {
            // print on stdout (native) or js console (web)
            App::print(format!("{:?}", evt));
            // exit when pressing escape
            match &evt {
                AppEvent::KeyUp(ev) if ev.code == ScanCode::Escape => {
                    App::exit();
                }
                _ => (),
            }
        }
    });
}
